### PR TITLE
o/snapstate: allow killing running snap apps during removal

### PR DESCRIPTION
**Pull request by ZeyadYasser**
This is needed in preparation for the new snap remove --terminate flag.

This PR ties all the pieces together through adding `kill-snap-apps` task along with the corresponding backend helper `KillSnapApps`.

This PR is part 3.2 of implementing force removal of snaps according to [SD174 - Force removal of snaps](https://docs.google.com/document/d/1ikDI-vPJO6YYFHOyFJ5hfK5L5Gv8voMeY8HJvxNWujw/edit#bookmark=id.ir3zex7innn) spec.

For context:
- Part 1: https://github.com/snapcore/snapd/pull/14126
- Part 2: https://github.com/snapcore/snapd/pull/14144
- Part 3.1: https://github.com/snapcore/snapd/pull/14160

### CODE REVIEW

**Comment by MiguelPires** [comment]
@ZeyadYasser I see this depends on a few open PRs, should I review now? If so, is the last commit the only relevant one for this PR?

**Comment by ZeyadYasser** [comment]
> should I review now?

Yes please. The other PRs are all approved so there shouldn't be major changes but I am holding off merging until after the release. It is totally fine if you prefer to wait after they are merged, this PR is not that urgent.

>  If so, is the last commit the only relevant one for this PR?

I highlighted the relevant commits in the PR description, currently they are https://github.com/snapcore/snapd/commit/5e60ca633b562eb57352033279c03ee183f601d3 and https://github.com/snapcore/snapd/commit/a795e0fccf7bd01af8d80e732570b9f73c40449b, I will keep an updated list in the description.

> There's a merge conflict btw

Thanks, I fixed it now

**Comment by ZeyadYasser** [comment]
@pedronis @MiguelPires PR is now rebased and ready for review

File: overlord/snapstate/backend/link.go
Comment (pedronis) [inline overlord/snapstate/backend/link.go:414]: would it make sense to track timings here?

**Review by pedronis — APPROVED** [review]
thank you, one question

File: overlord/snapstate/backend/link.go
Comment (ZeyadYasser) [reply overlord/snapstate/backend/link.go]: yes, similar to starting/stopping service, updated!

Comment (ZeyadYasser) [reply overlord/snapstate/backend/link.go]: Also thanks for pointing this out, as I found a nasty bug while adding timings where basically the kill task gets stuck due to locked state because I was using `NewTaskProgressAdapterUnlocked`.

I found that there was no need to lock the state for killing apps or stopping services so I unlocked the state altogether like what is being done `stop-snap-services`. https://github.com/snapcore/snapd/pull/14189/commits/ba67be0451a4ad10f3680d2edd9ae4df206b95db

File: overlord/snapstate/backend/link.go
Comment (MiguelPires) [inline overlord/snapstate/backend/link.go:423]: This will overwrite previous error values so it might end up notifying of failures but then returning nil

**Review by MiguelPires — COMMENTED** [review]
Looks good but quick comment

File: overlord/snapstate/backend/link.go
Comment (ZeyadYasser) [reply overlord/snapstate/backend/link.go:423]: I don't understand, which previous error values?

Comment (MiguelPires) [reply overlord/snapstate/backend/link.go:423]: Oh my bad, I think I mistook this as looping somehow (in my defence, it was late :p). Ignore me, of course this looks fine

**Review by MiguelPires — APPROVED** [review]
lgtm, thank you

**Comment by ZeyadYasser** [comment]
merged latest from master to include spread test fix

**Pull request merged**
